text/0000-cortex-m-targets.md

```diff
@@ -0,0 +1,608 @@
+instruction as all their floating point operations get lowered to intrinsics. This means that a
+program compiled with flags that enable FPU instructions *may* (\*) still end up using intrinsics
+instead of FPU instructions if the program calls a routine from e.g. `libcore.rlib` that performs a
+floating point operation.
+
+#### (\*)  Author note
+
+I tested the scenario described above. But, to my surprise, LLVM appears to be smart enough to
+convert an intrinsic back into a FPU instruction when CPU optimizations are enabled.
+
+Here's the program I tried:
+
+```
+$ cat src/main.rs
+```
+
+``` rust
+fn start() -> ! {
+    let x = 1f64;
+    let y = x.to_degrees();
+    
+    // ...
+}
+```
+
+`libcore.rlib` was compiled *without* CPU optimizations and, as seen below, its `to_degrees` routine
+uses the `__aeabi_dmul` intrinsic instead of FPU instructions.
+
+```
+$ arm-none-eabi-objdump -Cd $(rustc --print sysroot)/lib/rustlib/thumbv7em-none-eabi/lib/libcore-*.rlib
+(..)
+
+00000000 <core::f64::_$LT$impl$u20$core..num..Float$u20$for$u20$f64$GT$::to_degrees::he4eaab80e305699e>:
+   0:   b580            push    {r7, lr}
+   2:   f24c 12f8       movw    r2, #49656      ; 0xc1f8
+   6:   f24a 53dc       movw    r3, #42460      ; 0xa5dc
+   a:   f6c1 2263       movt    r2, #6755       ; 0x1a63
+   e:   f2c4 034c       movt    r3, #16460      ; 0x404c
+  12:   f7ff fffe       bl      0 <__aeabi_dmul>
+  16:   bd80            pop     {r7, pc}
+
+(..)
+```
+
+However, when I compiled the Cargo project with CPU optimizations enabled...
+
+````
+$ cargo rustc --target thumbv7em-none-eabi -- -C target-cpu=cortex-m7
+$ arm-none-eabi-objdump -Cd target/thumbv7em-none-eabi/debug/app
+(..)
+
+00000038 <core::f64::_$LT$impl$u20$core..num..Float$u20$for$u20$f64$GT$::to_degrees::he4eaab80e305699e>:
+  38:   b082            sub     sp, #8
+  3a:   ec41 0b10       vmov    d0, r0, r1
+  3e:   ed8d 0b00       vstr    d0, [sp]
+  42:   ed9f 1b05       vldr    d1, [pc, #20]   ; 58 <core::f64::_$LT$impl$u20$core..num..Float$u20$for$u20$f64$GT$::to_degrees::he4eaab80e305699e+0x20>
+  46:   ee20 0b01       vmul.f64        d0, d0, d1
+  4a:   ec51 0b10       vmov    r0, r1, d0
+  4e:   b002            add     sp, #8
+  50:   4770            bx      lr
+  52:   bf00            nop
+  54:   bf00            nop
+  56:   bf00            nop
+  58:   1a63c1f8        .word   0x1a63c1f8
+  5c:   404ca5dc        .word   0x404ca5dc
+
+(..)
+```
+
+:tada: LLVM updated the `to_degrees` function by replacing the `__aeabi_dmul` intrinsic with a more
+performant implementation that uses the `vmul.vf64` FPU instruction.
+
+However, **I'm** not sure if we can rely on LLVM **always** being capable of doing this.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+## Linker field
+
+Should we set the `linker` field of all the target specifications to `arm-none-eabi-gcc`?
+`arm-none-eabi-gcc` seems to be the most common (or maybe the only one?) bare-metal ARM toolchain.
```

Comment:
I think it may be best to use arm-none-eabi-ld instead.  The user would only require the binutils package instead of the entire GCC toolchain, and I don't see any need to automatically link in the default libs that arm-none-eabi-gcc implicity brings in.